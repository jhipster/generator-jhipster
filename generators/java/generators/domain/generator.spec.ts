import { basename, dirname, resolve } from 'path';
import { fileURLToPath } from 'url';
import { before, describe, expect, it } from 'esmocha';

import { shouldSupportFeatures, testBlueprintSupport } from '../../../../test/support/tests.js';
import { defaultHelpers as helpers, result } from '../../../../lib/testing/index.js';
import Generator from './index.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const generator = `${basename(resolve(__dirname, '../../'))}:${basename(__dirname)}`;

describe(`generator - ${generator}`, () => {
  shouldSupportFeatures(Generator);
  describe('blueprint support', () => testBlueprintSupport(generator));

  describe('with default config', () => {
    before(async () => {
      await helpers.runJHipster(generator).withJHipsterConfig({}, [
        {
          name: 'Foo',
          fields: [
            { fieldName: 'name', documentation: 'My Name', fieldType: 'String', fieldValidateRules: ['required'] },
            { fieldName: 'myEnum', fieldType: 'MyEnum', fieldValues: 'FRENCH,ENGLISH', fieldTypeDocumentation: 'Enum Doc' },
          ],
        },
        {
          name: 'Bar',
          documentation: 'Custom Bar',
          fields: [{ fieldName: 'name2', fieldType: 'String', fieldValidateRules: ['required'] }],
        },
      ]);
    });

    it('should match files snapshot', () => {
      expect(result.getStateSnapshot()).toMatchSnapshot();
    });

    it('should generate entities containing jakarta', () => {
      result.assertFileContent('src/main/java/com/mycompany/myapp/domain/Foo.java', 'jakarta');
    });

    it('should generate javadocs', () => {
      result.assertFileContent('src/main/java/com/mycompany/myapp/domain/Foo.java', '* A Foo');
      result.assertFileContent('src/main/java/com/mycompany/myapp/domain/Foo.java', '* My Name');
      result.assertFileContent('src/main/java/com/mycompany/myapp/domain/Bar.java', '* Custom Bar');
    });

    it('should generate openapi @Schema', () => {
      result.assertFileContent('src/main/java/com/mycompany/myapp/domain/Bar.java', '@Schema(description = "Custom Bar")');
    });

    it('should write enum files', () => {
      result.assertFile('src/main/java/com/mycompany/myapp/domain/enumeration/MyEnum.java');
      expect(Object.keys(result.getStateSnapshot('**/enumeration/**')).length).toBe(2);
    });

    it('should generate enum javadoc', () => {
      result.assertFileContent('src/main/java/com/mycompany/myapp/domain/enumeration/MyEnum.java', '* Enum Doc');
    });

    it('should have options defaults set', () => {
      const generator: any = result.generator;
      expect(generator.generateEntities).toBe(true);
      expect(generator.generateEnums).toBe(true);
      expect(generator.useJakartaValidation).toBe(true);
    });
  });

  describe('with jakarta and enums disabled', () => {
    before(async () => {
      await helpers
        .runJHipster(generator)
        .withJHipsterConfig({}, [
          {
            name: 'Foo',
            fields: [
              { fieldName: 'name', fieldType: 'String', fieldValidateRules: ['required'] },
              { fieldName: 'myEnum', fieldType: 'MyEnum', fieldValues: 'FRENCH,ENGLISH' },
            ],
          },
        ])
        .withOptions({ useJakartaValidation: false, generateEnums: false });
    });

    it('should generate entities not containing jakarta', () => {
      result.assertNoFileContent('src/main/java/com/mycompany/myapp/domain/Foo.java', 'jakarta');
    });

    it('should not write enum files', () => {
      expect(Object.keys(result.getStateSnapshot('**/enumeration/**')).length).toBe(0);
    });
  });

  describe('with entities disabled', () => {
    before(async () => {
      await helpers
        .runJHipster(generator)
        .withJHipsterConfig({}, [
          {
            name: 'Foo',
            fields: [
              { fieldName: 'name', fieldType: 'String', fieldValidateRules: ['required'] },
              { fieldName: 'myEnum', fieldType: 'MyEnum', fieldValues: 'FRENCH,ENGLISH' },
            ],
          },
        ])
        .withOptions({ generateEntities: false });
    });

    it('should not contain jakarta', () => {
      result.assertNoFile('src/main/java/com/mycompany/myapp/domain/Foo.java');
    });
  });

  describe('with custom properties values', () => {
    before(async () => {
      await helpers
        .runJHipster(generator)
        .withJHipsterConfig({})
        .onGenerator((generator: any) => {
          generator.generateEntities = false;
          generator.generateEnums = false;
          generator.useJakartaValidation = false;
        });
    });

    it('should not override custom values', () => {
      const generator: any = result.generator;
      expect(generator.generateEntities).toBe(false);
      expect(generator.generateEnums).toBe(false);
      expect(generator.useJakartaValidation).toBe(false);
    });
  });
});
