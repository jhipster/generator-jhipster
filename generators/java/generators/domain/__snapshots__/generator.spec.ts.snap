// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`generator - java:domain with default config should match files snapshot 1`] = `
{
  ".jhipster/Bar.json": {
    "stateCleared": "modified",
  },
  ".jhipster/Foo.json": {
    "stateCleared": "modified",
  },
  ".yo-rc.json": {
    "stateCleared": "modified",
  },
  "src/main/java/com/mycompany/myapp/domain/Authority.java": {
    "stateCleared": "modified",
  },
  "src/main/java/com/mycompany/myapp/domain/Bar.java": {
    "stateCleared": "modified",
  },
  "src/main/java/com/mycompany/myapp/domain/Foo.java": {
    "stateCleared": "modified",
  },
  "src/main/java/com/mycompany/myapp/domain/enumeration/MyEnum.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/AssertUtils.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/AuthorityAsserts.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/AuthorityTest.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/AuthorityTestSamples.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/BarAsserts.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/BarTest.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/BarTestSamples.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/FooAsserts.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/FooTest.java": {
    "stateCleared": "modified",
  },
  "src/test/java/com/mycompany/myapp/domain/FooTestSamples.java": {
    "stateCleared": "modified",
  },
}
`;
